scheduling-utils/src/handshake/client.rs

```diff
@@ -0,0 +1,140 @@
+        time::Duration,
+    },
+    thiserror::Error,
+};
+
+type ShaqError = shaq::error::Error;
+
+/// The maximum size in bytes of the control message containing the queues assuming [`MAX_WORKERS`]
+/// is respected.
+///
+/// There are 2 global queues (tpu and progress) and 2 queues per worker (1 each way with the
+/// external scheduler). Each FD is 4 bytes in memory, so the size of our control message (excluding
+/// the header) is `(global_queues + workers * 2) * 4`.
+const CMSG_MAX_SIZE: usize = (2 + MAX_WORKERS * 2) * 4;
+
+/// Connects to the scheduler server on the given IPC path.
+///
+/// # Timeout
+///
+/// Timeout is enforced at the syscall level. In the typical case, this function will do two
+/// syscalls, one to send the logon message and one to receive the response. However, if for
+/// whatever reason the OS does not accept 1024 bytes in a single syscall, then multiple writes
+/// could be needed. As such this timeout is meant to guard against a broken server but not
+/// necessarily ensure this function always returns before the timeout (this is somewhat in line
+/// with typical timeouts because you have no guarantee of being rescheduled).
+pub fn connect(
+    path: impl AsRef<Path>,
+    logon: &ClientLogon,
+    timeout: Duration,
+) -> Result<ClientSession, ConnectError> {
+    connect_path(path.as_ref(), logon, timeout)
+}
+
+fn connect_path(
+    path: &Path,
+    logon: &ClientLogon,
+    timeout: Duration,
+) -> Result<ClientSession, ConnectError> {
+    let mut stream = UnixStream::connect(path)?;
+    stream.set_read_timeout(Some(timeout))?;
+    stream.set_write_timeout(Some(timeout))?;
+
+    // Send the logon message.
+    let mut buf = [0; 1024];
+    buf[..ClientLogon::LEN].copy_from_slice(bytemuck::bytes_of(logon));
+    stream.write_all(&buf)?;
+
+    // Receive the requested FDs.
+    let mut buf = [0; 1024];
+    let mut iov = [IoSliceMut::new(&mut buf)];
+    // SAFETY: CMSG_LEN is always safe (const expression).
+    let mut cmsgs = [0u8; unsafe { CMSG_LEN(CMSG_MAX_SIZE as u32) as usize }];
+    let msg = socket::recvmsg::<UnixAddr>(
+        stream.as_raw_fd(),
+        &mut iov,
+        Some(&mut cmsgs),
+        MsgFlags::empty(),
+    )?;
+
+    // Check for failure.
+    let buf = msg.iovs().next().unwrap();
+    if buf[0] == LOGON_FAILURE {
+        let reason_len = usize::from(buf[1]);
+        #[allow(clippy::arithmetic_side_effects)]
+        let reason = std::str::from_utf8(&buf[2..2 + reason_len]).unwrap();
+
+        return Err(ConnectError::Rejected(reason.to_string()));
+    }
+
+    // Extract FDs.
+    let mut cmsgs = msg.cmsgs().unwrap();
+    let fds = match cmsgs.next() {
+        Some(ControlMessageOwned::ScmRights(fds)) => fds,
+        Some(msg) => panic!("Unexpected; msg={msg:?}"),
+        None => panic!(),
+    };
+    assert!(fds.len() % 2 == 0 && fds.len() >= 4);
+    let workers = (fds.len().wrapping_sub(2)) / 2;
+    assert_eq!(workers, logon.worker_count);
```

Comment:
let's return errors instead of asserting these things.

This can happen if there's a mismatch in expectations of fd count between server/client, right?

Reply:
It's a fatal protocol violation, I don't see how the client can recover from this without a deploying a new binary?

Reply:
Hmm, I personally much prefer panics. If you run your services with:
```
RUST_LIB_BACKTRACE=0
RUST_BACKTRACE=1
```

You get backtraces on panics but not on errors so you get backtraces in prod without really paying a performance penalty (unless some code is chronically catching unwinds). Even if the user unwraps the error they won't really know what line threw it (unless the error variant is used in a single place). Generally I've found panics to be a better devex (unless the error is recoverable then ofc catch unwind is horrific).

Reply:
sorry to be clear, if this were an internal error i'd agree with you - just panic.
this is something that is dependent on server & client matching expectations, so is better to return an error imo.